openedx/openedx-platform
# feat: Add REST endpoints to update Components in a Collections [FC-0062]

## Description

Adds a REST endpoint (patch, delete) for updating the Components in a v2 Content Library's Collection.
Adding/removing components raises a [CONTENT_OBJECT_TAGS_CHANGED](https://github.com/openedx/openedx-events/blob/main/openedx_events/content_authoring/signals.py#L207) event, because we will index the ContentObject's collections as a type of "tag".

Note: We'll be renaming `CONTENT_OBJECT_TAGS_CHANGED` to `CONTENT_OBJECT_ASSOCIATIONS_CHANGED` as part of https://github.com/openedx/modular-learning/issues/230

This PR adds to the features available to Course Authors of v2 Libraries, which aren't recommended for production use yet.

## Supporting information

Part of : https://github.com/openedx/modular-learning/issues/227
Based on: https://github.com/openedx/edx-platform/pull/35321
Depends on:
- https://github.com/openedx/openedx-learning/pull/216
- https://github.com/openedx/openedx-events/pull/386

Private-ref: [FAL-3784](https://tasks.opencraft.com/browse/FAL-3784)

## Testing instructions

Please check that the unit tests added here cover most cases.

Manual testing:
1. Checkout this branch in your local dev environment
1. Make sure you have https://github.com/openedx/openedx-events/pull/386 and  https://github.com/openedx/openedx-learning/pull/216 installed and running locally 
1. Create a V2 library, e.g `lib:SampleTaxonomyOrg1:AL1`
2. Add some blocks to that library.
1. In an external HTTP client (like curl/postman or similar) make the following requests (make sure you have the correct headers for authentication, you can copy them from the network tab in your browser), replace the library key in the URL with your actual library key:
    1. POST http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/
       1. Payload: `{"title":"Collection 1", "description": "First Collection"}`
       1. This should return the newly created collection; note the ID and `modified` dates.
    1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/blocks/
       1. Should return the list of blocks added to this library
       1. Note their usage key values for the steps below.
    1. PATCH http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/1/contents/
        1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
        1. Should return `{"count": <number of components in list>}`
        1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
    1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/1/
        1. The `modified` timestamp should be updated.
    1. DELETE http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/1/contents/
        1. Payload: `{"usage_keys": [ <list of usage keys from the library blocks> ]`
        1. Should return `{"count": <number of components in list>}`
        1. Check the logs and confirm that the event was emitted for each component, triggering an "upsert" into Meilisearch.
    1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/1/
        1. The `modified` timestamp should be updated.
    1. GET http://studio.local.edly.io:8001/api/libraries/v2/lib:SampleTaxonomyOrg1:AL1/collections/1/contents/ - will raise a 405, as an object's collections will be fetched from Meilisearch.
